### PR TITLE
Fix a bug in rotational periodic boundary conditions

### PR DESCRIPTION
**Pull request by GuySten**
# Description

This PR simplify and fix logic of periodic boundary conditions.
It prevents particles getting lost when the periodic surfaces have normals in any direction.
I also added tests that check that the code run correctly regardless of the periodic plane normal direction.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] ~~I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)~~
- [x] ~~I have made corresponding changes to the documentation (if applicable)~~
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for this fix and simplification! The one thing that still doesn't seem quite right is the warning about the angle not evenly subdividing 360. Depending on the orientation of the surfaces and whether positive/negative half-spaces are used to define the region, sometimes that warning is shown even when the angle is fine. To illustrate that,  I've put together this example with two planes that are created using `Plane.from_points`, where each set of three points is randomly permuted. The region being plotted is the same in each case but some permutations of the points end up with the warning and some don't.
```python
import openmc
from math import cos, sin, pi, radians
import numpy as np
import random

start = 0.
degrees = 45.
ang1 = radians(start)
ang2 = radians(start + degrees)

p1_points = [(0., 0., 0.), (cos(ang1), sin(ang1), 0.), (0., 0., 1.)]
p2_points = [(0., 0., 0.), (cos(ang2), sin(ang2), 0.), (0., 0., 1.)]
random.shuffle(p1_points)
random.shuffle(p2_points)

p1 = openmc.Plane.from_points(*p1_points, boundary_type='periodic')
p2 = openmc.Plane.from_points(*p2_points, boundary_type='periodic')
p1.periodic_surface = p2
zcyl = openmc.ZCylinder(r=5., boundary_type='vacuum')

# Figure out which side of planes to use based on a point in the middle
ang_mid = radians(start + degrees/2.)
mid_point = (cos(ang_mid), sin(ang_mid), 0.)
r1 = -p1 if mid_point in -p1 else +p1
r2 = -p2 if mid_point in -p2 else +p2

(r1 & r2 & -zcyl).plot()
```

**Comment by GuySten** [comment]
In my testing, the normalization of the angle range to [-PI,PI] fix the warning problem.

**Review by paulromano — COMMENTED** [review]
Thanks for the update. The update seems to have fixed the warning, but now cases that are 90 or 120 degrees rotationally periodic result in lost particles. I added a test to check for this. Can you look into this further?

**Comment by GuySten** [comment]
I had to compute the periodic surfaces senses to compute the angle using the canonical normal direction.

File: src/surface.cpp [hunk 1]
```diff
@@ -1276,6 +1279,28 @@ void read_surfaces(pugi::xml_node node)
     }
   }
 
+  // Fill the senses map for periodic surfaces
+  auto v = node.children("cell");
+  auto n_periodic = periodic_sense_map.size();
+  for (auto it = v.begin(); (it != v.end()) && (n_periodic > 0); ++it) {
+    pugi::xml_node cell_node = *it;
```
Comment (paulromano) [inline src/surface.cpp]: I understand the need to get the sense information from cell region definitions, but the code you've added here will result in us iterating over all cells (and reading their regions) twice, which is not ideal. Instead, I would suggest deferring the creation of the periodic BC objects until after cells have been read so that we can get the sense information without having to redundantly parse the XML.

**Review by paulromano — CHANGES_REQUESTED** [review]

File: src/surface.cpp (hunk 1 above)
Comment (GuySten) [reply src/surface.cpp]: Done.

**Comment by lewisgross1296** [comment]
@GuySten thanks for this! Could you explain exactly what the problem was and what the solution was? 

I think I might have experienced an error related to this. After using the new periodic BC, everything seemed to be working, but during one of my transport steps (as part of a Cardinal multiphysics simulation where OpenMC runs many times), the below happened
```
 Running OpenMC with 50000 particles per batch...
Executing filter editors...
Executing tally editors...
 Maximum neutron transport energy: 20000000 eV for He3
 Initializing source particles...

 ====================>     K EIGENVALUE SIMULATION     <====================

  Bat./Gen.      k            Average k
  =========   ========   ====================
        1/1    0.72088
...
       28/1    1.00152
       29/1    1.01043
       30/1   -nan    
```
This happened after four successful OpenMC runs, so it must be caused by a rare event. The error I got was 
```
 WARNING: Particle 20017 underwent maximum number of events.
 ERROR: No fission sites banked on MPI rank 31
 ERROR: No fission sites banked on MPI rank 23
 ERROR: No fission sites banked on MPI rank 27
 ERROR: No fission sites banked on MPI rank 34
 ERROR: No fission sites banked on MPI rank 10
 ERROR: No fission sites banked on MPI rank 41
 ERROR: No fission sites banked on MPI rank 1
 ERROR: No fission sites banked on MPI rank 35
 ERROR: No fission sites banked on MPI rank 9
 ERROR: No fission sites banked on MPI rank 15
 ERROR: No fission sites banked on MPI rank 19
 ERROR: No fission sites banked on MPI rank 44
 ERROR: No fission sites banked on MPI rank 38
 ERROR: No fission sites banked on MPI rank 13
 ERROR: No fission sites banked on MPI rank 28
 ERROR: No fission sites banked on MPI rank 11
 ERROR: No fission sites banked on MPI rank 14
 ERROR: No fission sites banked on MPI rank 8
 ERROR: No fission sites banked on MPI rank 26
 ERROR: No fission sites banked on MPI rank 12
 ERROR: No fission sites banked on MPI rank 42
 ERROR: No fission sites banked on MPI rank 5
 ERROR: No fission sites banked on MPI rank 21
 ERROR: No fission sites banked on MPI rank 20
 ERROR: No fission sites banked on MPI rank 29
 ERROR: No fission sites banked on MPI rank 22
 ERROR: No fission sites banked on MPI rank 18
 ERROR: No fission sites banked on MPI rank 7
 ERROR: No fission sites banked on MPI rank 25
 ERROR: No fission sites banked on MPI rank 46
 ERROR: No fission sites banked on MPI rank 45
 ERROR: No fission sites banked on MPI rank 3
 ERROR: No fission sites banked on MPI rank 6
 ERROR: No fission sites banked on MPI rank 24
 ERROR: No fission sites banked on MPI rank 47
 ERROR: No fission sites banked on MPI rank 30
 ERROR: No fission sites banked on MPI rank 37
 ERROR: No fission sites banked on MPI rank 33
 ERROR: No fission sites banked on MPI rank 40
 ERROR: No fission sites banked on MPI rank 39
 ERROR: No fission sites banked on MPI rank 16
 ERROR: No fission sites banked on MPI rank 36
 ERROR: No fission sites banked on MPI rank 4
 ERROR: No fission sites banked on MPI rank 0
 ERROR: No fission sites banked on MPI rank 32
 ERROR: No fission sites banked on MPI rank 43
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 26 in communicator MPI_COMM_WORLD
with errorcode -1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
 ERROR: No fission sites banked on MPI rank 2
 ERROR: No fission sites banked on MPI rank 17
```
Does this seem similar to the problems you noticed in your testing?

**Comment by GuySten** [comment]
The only problems I encountered are missing particles.
 
The problem was that If you have two intersecting planes the angle between them is not well defined (there are two possible angles).
The angle between them is well defined only if they are halfspaces.
And you have to know how the halfspace is used in the geometry to get the correct orientation of the halfspace to compute the correct angle.
The rest is just math.

**Comment by lewisgross1296** [comment]
I see, yea in this case I do get the warning about 299.999 degrees not evenly dividing into 360, so in my case it's measuring the other angle. 

Do you think the `-nan` I encountered would be fixed by the changes in this PR? I thought from the title of this PR and reply on [discource](https://openmc.discourse.group/t/flip-normal-and-physics-for-periodic-problems/6148) that this PR might to address the `-nan` as well as the angle warning. 

Though, I wonder if
```
 WARNING: Particle 20017 underwent maximum number of events.
```
might be related to an issue with the current implementation of PeriodicBCs. Perhaps it's a separate problem? And maybe it's causing the no fission sites banked because all the other successful runs had enough fission sites per MPI process

**Comment by GuySten** [comment]
I don't know but you can install this branch and check if that solves the problem.

**Comment by lewisgross1296** [comment]
Since my issue was happening within a Cardinal simulation, and after multiple OpenMC solves on a computationally expensive run, it's not super straightforward to just build Cardinal with your OpenMC change and test to see if it fixes it. (It might take order of days to find out if this worked). However, I'm going to do some testing to see if I can prove my issue is unrelated / a coincidence. From what I can tell from your comments/source code changes, this PR is only aimed to fix the warning, and is unrelated to my `-nan` issue.

Back to the original discourse forum post I made (re-sharing the image here)
image
Should we add a test that using `flip_normal` method on a surface involved with a periodic boundary condition does not change the eigenvalue? I think there may be something else strange with my model causing the `-nan` but at the very least, I don't think using the `flip_normal` method on a surface should change the actual eigenvalue as I observed.

File: src/boundary_condition.cpp [hunk 2]
```diff
@@ -160,7 +160,7 @@ void TranslationalPeriodicBC::handle_particle(
 
 RotationalPeriodicBC::RotationalPeriodicBC(
   int i_surf, int j_surf, PeriodicAxis axis)
-  : PeriodicBC(i_surf, j_surf)
+  : PeriodicBC(std::abs(i_surf) - 1, std::abs(j_surf) - 1)
```
Comment (lewisgross1296) [inline src/boundary_condition.cpp:163]: Can you explain where the `-1` comes from? This seems like a significant change. Tagging @gonuke who was also curious.

File: src/boundary_condition.cpp [hunk 3]
```diff
@@ -173,14 +173,9 @@ RotationalPeriodicBC::RotationalPeriodicBC(
...
-    // for a right handed coordinate system, z should be the independent axis
-    // but this would cause the y-rotation case to be different than the other
-    // two. using a left handed coordinate system and a negative rotation the
-    // compute angle and rotation matrix behavior mimics that of the x and z
-    // cases
     zero_axis_idx_ = 1; // y component of plane must be zero
-    axis_1_idx_ = 0;    // x component independent
-    axis_2_idx_ = 2;    // z component dependent
+    axis_1_idx_ = 2;    // z component independent
+    axis_2_idx_ = 0;    // x component dependent
```
Comment (lewisgross1296) [inline src/boundary_condition.cpp:178]: @GuySten I think lines 176 - 183 were originally your suggestion in my [PR](https://github.com/openmc-dev/openmc/pull/3591)  (i.e. switching the independent and dependent axis for the y-case to make them confirm to the other cases). Can you explain why this is now being changed / how making this change does not break the assumptions of the rotation matrix code when computing `new_r` and `new_u`?

File: src/boundary_condition.cpp (hunk 2 above)
Comment (GuySten) [reply src/boundary_condition.cpp:163]: Because I wanted signed surface indices in the RotationalPeriodicBC class, I had to use the convention that the input indices are 1-indexed. So this code convert signed 1-indexed surfaces to unsigned 0-indexed surfaces.
The surfaces signs are used solely for computing the correct angle in this scope and afterwards they are dropped.

File: src/boundary_condition.cpp (hunk 3 above)
Comment (GuySten) [reply src/boundary_condition.cpp:178]: Transpose of a rotation matrix with angle theta is equivalent to a rotation matrix with angle -theta. In the previous PR we could not deduce the right rotation angle because we did not use the information about the surface orientation. So we effectively guessed it to pass the test cases. There was no reason to special case the 'y' rotation at all. It is not inherently special. The reason we had to write an algorithm that behaved differently in the y case had to do with the limited test cases.
When paul added some test cases and I added more no algorithm with any special cases could pass all tests without using the surfaces orientation.

**Comment by lewisgross1296** [comment]
Using `openmc.Geometry.merge_surfaces = True` ended up removing lines from my XML. I've now had one case get past the `-nan` where it originally failed, so some duplicated surfaces could have caused my problem. This suggests my `-nan`  issue is specific to my model and shouldn't block anything here. Sorry, I originally couldn't tell if this PR was meant to fix both the angle and the nan from your response on discourse. Thanks for this contribution!

**Review by paulromano — APPROVED** [review]
Thanks for working through this bug @GuySten!
